### PR TITLE
PHP-1407: Implement verify_peer_name for PHP < 5.6.0 

### PR DESCRIPTION
**Pull request by bjori**

### CODE REVIEW

File: io_stream.c [hunk 1]
```diff
@@ -141,6 +185,14 @@ void* php_mongo_io_stream_connect(mongo_con_manager *manager, mongo_server_def *
 				return NULL;
 			}
 		} else {
+#if PHP_VERSION_ID < 50600
+			if (php_mongo_verify_hostname(manager, stream, server TSRMLS_CC) == FAILURE) {
```
Comment (jmikola) [inline io_stream.c]: This is prone to returning `FAILURE` even if the user isn't requesting host name validation via the `peer_certificate` SSL stream context option.

File: io_stream.c [hunk 2]
```diff
@@ -89,6 +120,15 @@ void* php_mongo_io_stream_connect(mongo_con_manager *manager, mongo_server_def *
 	} else {
 		mongo_manager_log(manager, MLOG_CON, MLOG_FINE, "Connecting to %s (%s) without connection timeout (default_socket_timeout will be used)", dsn, hash);
 	}
+#if PHP_VERSION_ID < 50600
+	/* Capture the server certificate, if SSL is enabled, so we can do hostname verification */
```
Comment (jmikola) [inline io_stream.c]: Do you mean for this to *always* happen in pre-5.6, regardless of the context options?

Comment (bjori) [inline io_stream.c]: fixed.. it was only supposed to run when verify_peer_name was provided

File: contrib/php-ssl.c [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+         **/
+        if ( strcasecmp(wildcard + 1, subjectname + subject_len - suffix_len) == 0 &&
+            memchr(subjectname + prefix_len, '.', subject_len - suffix_len - prefix_len) == NULL) {
+			return SUCCESS;
+		}
+    }
+
+    return FAILURE;
+}
+/* }}} */
```
Comment (derickr) [inline contrib/php-ssl.c]: This is a spurious marker, there is none on the other side.

File: contrib/php-ssl.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  |          Chris Wright <daverandom@php.net>                           |
+  +----------------------------------------------------------------------+
+*/
+
+#ifndef MONGO_CONTRIB_SSL_H
+#define MONGO_CONTRIB_SSL_H
+
+#include "php.h"
+#include <openssl/x509.h>
+#include <openssl/x509v3.h>
```
Comment (derickr) [inline contrib/php-ssl.h]: Doesn't this only work if the openssl headers are actually installed? Config.m4 only checks for SASL, and not openssl.

File: io_stream.c [hunk 5]
```diff
@@ -116,7 +158,11 @@ void* php_mongo_io_stream_connect(mongo_con_manager *manager, mongo_server_def *
 
 		zend_replace_error_handling(EH_THROW, mongo_ce_ConnectionException, &error_handler TSRMLS_CC);
 
+#if PHP_VERSION_ID >= 50600 && PHP_VERSION_ID < 50607
```
Comment (derickr) [inline io_stream.c]: Perhaps add a link to the bug report / PHP Commit here? - or at least an explanation why 5.6.0-5.6.7 are different in a comment?

File: io_stream.c [hunk 6]
```diff
@@ -141,6 +187,15 @@ void* php_mongo_io_stream_connect(mongo_con_manager *manager, mongo_server_def *
 				return NULL;
 			}
 		} else {
+#if PHP_VERSION_ID < 50600
+			if (verify && zend_is_true(*verify)) {
+				if (php_mongo_verify_hostname(manager, stream, server TSRMLS_CC) == FAILURE) {
+					*error_message = strdup("Hostname doesn't match");
```
Comment (derickr) [inline io_stream.c]: This message should include the expected hostname, and the given hostname - and possibly also that this is due to the SSL certificate.

**Comment by derickr** [comment]
And lastly... any thoughts on how to test this? Right now, this PR doesn't have any tests.

File: contrib/php-ssl.h (hunk 4 above)
Comment (bjori) [inline contrib/php-ssl.h]: Good point.
I guess we should wrap all this in HAVE_OPENSSL_EXT

File: io_stream.c (hunk 5 above)
Comment (bjori) [inline io_stream.c]: makes sense

**Comment by bjori** [comment]
Its really difficult to test for on an unknown setup.
Bundling the certificates wouldn't be enough as (per the feature) it would require your host to have specific name.
We can test this in phongo as we spin up dedicated vms that we have full control over, including hostnames, but I can't think of a way to automate a test for this here

**Comment by bjori** [comment]
All comments should been address now.
Closing this PR however in favor of https://github.com/mongodb/mongo-php-driver/pull/836 which includes this one and addresses more related issues

**Pull request closed without merging**